src/tools/clippy/clippy_utils/src/ty.rs

```diff
@@ -128,9 +129,10 @@ pub fn implements_trait<'tcx>(
         return false;
     }
     let ty_params = cx.tcx.mk_substs(ty_params.iter());
-    cx.tcx
-        .type_implements_trait((trait_id, ty, ty_params, cx.param_env))
+    cx.tcx.infer_ctxt().enter(|infcx|
```

Comment:
Can you modify the `ty.has_infer_types` above to `ty.needs_infer`, which is a more complete fix? (Actually, I think I'd prefer an assertion, but that's more of a matter for @rust-lang/clippy to decide -- I would be surprised though if inference variables make their way here)

For that matter, I think the function should have a comment explaining that it returns false in the case of inference variables, as that is not obvious (also not obviously correct)

Reply:
The `ty.has_infer_types()` was added as a fix for a ICE: https://github.com/rust-lang/rust/pull/72775

Reply:
Makes sense -- `needs_infer` would probably avoid *more* ICEs :)

Reply:
Note though @flip1995 that there's probably a "better fix" that just makes the code *work*

Reply:
Oh, I see, but the ICE in question was https://github.com/rust-lang/rust/issues/72766 and is not part of clippy, right?

Reply:
it might well be that this can be an `assert!` in the clippy code

Reply:
I really don't know if this is required in Clippy, I've never tested it without it since it was added. We call this utils function in many places, so it's likely that this actually prevents an ICE, even if it is just there as a precaution, so that this function can just be used without checking preconditions.